src/internal/pjs/api_server.go

```diff
@@ -184,8 +171,108 @@ func (a *apiServer) ProcessQueue(srv pjs.API_ProcessQueueServer) (retErr error)
 				}
 				return nil
 			}); err != nil {
-				return err
+				return errors.Wrap(err, "with tx")
+			}
+		}
+	}
+}
+
+func (a *apiServer) WalkJob(req *pjs.WalkJobRequest, srv pjs.API_WalkJobServer) (err error) {
+	ctx, done := log.SpanContext(srv.Context(), "walkJob")
+	defer done(log.Errorp(&err))
+
+	// handle job context and request validation.
+	var id pjsdb.JobID
+	if req.Context != "" {
+		id, err = a.resolveJobCtx(ctx, req.Context)
+		if err != nil {
+			return errors.Wrap(err, "resolve job ctx")
+		}
+	} else { //nolint:staticcheck
+		// TODO(PJS): do auth.
+	}
+	// TODO(PJS): remove this once auth is implemented.
+	reqID := pjsdb.JobID(req.Job.Id)
+	if id != reqID {
+		log.Error(ctx, "job context token does not match requested job.", zap.Int64("request.Job.ID", req.Job.Id))
+		id = reqID // defaulting to this for testing until auth is implemented.
+	}
+
+	// walk and stream back results.
+	var jobs []pjsdb.Job
+	if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+		jobs, err = pjsdb.WalkJob(ctx, sqlTx, id, pjsdb.WalkAlgorithm(req.Algorithm.Number()), req.MaxDepth)
+		return errors.Wrap(err, "walk job in pjsdb")
+	}, dbutil.WithReadOnly()); err != nil {
+		return errors.Wrap(err, "with tx")
+	}
+	for i, job := range jobs {
+		jobInfo, err := toJobInfo(job)
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf("to job info, iteration=%d/%d", i, len(jobs)))
+		}
+		resp := &pjs.ListJobResponse{
+			Id:   jobInfo.Job,
+			Info: jobInfo,
+			Details: &pjs.JobInfoDetails{
+				JobInfo: jobInfo,
+			},
+		}
+		if err := srv.Send(resp); err != nil {
+			return errors.Wrap(err, fmt.Sprintf("send, iteration=%d/%d", i, len(jobs)))
+		}
+	}
+	return nil
+}
+
+func (a *apiServer) resolveJobCtx(ctx context.Context, jobCtx string) (id pjsdb.JobID, err error) {
+	token, err := pjsdb.JobContextTokenFromString(jobCtx)
```

Comment:
I took a look at the failed bazel test because I want to reuse your fullBinaryTree and resolveContext method 😄.
I think the problem comes from this line. If you replace JobContextTokenFromString with JobContextTokenFromHex, that failed test will pass. This is because in api_server line 146, the token is already encoded into hex format.

Reply:
Thanks a bunch Muyang!